app/controllers/sessions_controller.rb

```diff
@@ -3,14 +3,24 @@ def new
   end
 
   def create
-    user = User.find_by_email(params[:session][:email])
-    if user && user.authenticate(params[:session][:password])
-      session[:user_id] = user.id
-      flash[:success] = "Welcome, #{user.full_name}."
+    if params[:session]
+      user = User.find_by_email(params[:session][:email])
+      if user && user.authenticate(params[:session][:password])
+        session[:user_id] = user.id
+        flash[:success] = "Welcome, #{user.full_name}."
+      else
+        flash[:danger] = "Invalid login."
+      end
```

Comment:
These conditionals are hard to follow. We should abstract them into their own private methods.